nemo/collections/llm/gpt/model/llama4_utils.py

```diff
@@ -0,0 +1,154 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from copy import deepcopy
+from typing import TYPE_CHECKING, Optional
+
+import torch
+from megatron.core.inference.contexts import BaseInferenceContext
+from megatron.core.models.gpt.gpt_layer_specs import get_gpt_decoder_block_spec
+from megatron.core.packed_seq_params import PackedSeqParams
+
+try:
+    from megatron.core.transformer.spec_utils import ModuleSpec
+    from megatron.core.transformer.transformer_layer import TransformerLayer as MCoreTransformerLayer
+
+    HAVE_TE = True
+except ImportError:
+    HAVE_TE = False
+
+if TYPE_CHECKING:
+    pass
+
+
+def get_llama4_layer_spec(config: "Llama4Config") -> ModuleSpec:
+    """Get llama4 layer spec"""
+
+    from megatron.core.transformer.transformer_layer import get_transformer_layer_offset
+
+    # Use decoder_block_spec: set layer_specs as a list of individual layer specs
+    llama4_layer_spec = get_gpt_decoder_block_spec(config, use_transformer_engine=HAVE_TE)
+
+    updated_layer_specs = []
+    offset = get_transformer_layer_offset(config)
+    for idx, layer_spec in enumerate(llama4_layer_spec.layer_specs):
+        layer_no = idx + offset
+        updated_layer_spec = deepcopy(layer_spec)
+
+        updated_layer_spec.module = Llama4TransformerLayer
+        is_nope_layer = config.nope_layer_interval is not None and (layer_no + 1) % config.nope_layer_interval == 0
+        updated_layer_spec.params = {
+            'is_nope_layer': is_nope_layer,
+            'attention_chunk_size': config.attention_chunk_size,
+        }
+        if config.qk_l2_norm and not is_nope_layer:
+            # Use QK Norm
+            updated_layer_spec.submodules.self_attention.submodules.q_layernorm = L2Norm
+            updated_layer_spec.submodules.self_attention.submodules.k_layernorm = L2Norm
+        else:
+            updated_layer_spec.submodules.self_attention.submodules.q_layernorm = None
+            updated_layer_spec.submodules.self_attention.submodules.k_layernorm = None
+        updated_layer_specs.append(updated_layer_spec)
+
+    llama4_layer_spec.layer_specs = updated_layer_specs
+    return llama4_layer_spec
+
+
+class Llama4TransformerLayer(MCoreTransformerLayer):
+    """Updated Transformer Layer to enable skip rope in some layers"""
+
+    def __init__(self, is_nope_layer=False, attention_chunk_size=8192, *args, **kwargs):
+        self.is_nope_layer = is_nope_layer
+        self.attention_chunk_size = attention_chunk_size
+        super(Llama4TransformerLayer, self).__init__(*args, **kwargs)
+
+    def _forward_attention(
+        self,
+        hidden_states: Tensor,
+        attention_mask: Optional[Tensor] = None,
```

Comment:
I used this branch and got the following error. this is missing an import statement above
```
1: NameError: name 'Tensor' is not defined
 0:     hidden_states: Tensor,
 ```

Reply:
fixed, something got lost during merging